stl/inc/memory

```diff
@@ -1199,6 +1199,24 @@ private:
     friend shared_ptr<_Ty0> allocate_shared(const _Alloc& _Al_arg, _Types&&... _Args);
 #endif // !_HAS_CXX20
 
+// Should this be in a separate ifdef _HAS_CXX20 control block since it is a
+// different feature?
+#if _HAS_CXX20
+    template <class _Ty0>
+    friend enable_if_t<!is_unbounded_array_v<_Ty0>, shared_ptr<_Ty0>> make_shared_for_overwrite();
+
+    template <class _Ty0, class _Alloc>
+    friend enable_if_t<!is_unbounded_array_v<_Ty0>, shared_ptr<_Ty0>> allocate_shared_for_overwrite(
+        const _Alloc& _Al_arg);
+
+    template <class _Ty0>
+    friend enable_if_t<is_unbounded_array_v<_Ty0>, shared_ptr<_Ty0>> make_shared_for_overwrite(size_t _Count);
+
+    template <class _Ty0, class _Alloc>
+    friend enable_if_t<is_unbounded_array_v<_Ty0>, shared_ptr<_Ty0>> allocate_shared_for_overwrite(
+        const _Alloc& _Al_arg, size_t _Count);

```

Comment:
This is contrary to the "usual" way `enable_if` is used in the code. I would expect something along those lines:
```suggestion
    template <class _Ty0, enable_if_t<!is_unbounded_array_v<_Ty0>, int> = 0>
    friend shared_ptr<_Ty0> make_shared_for_overwrite();

    template <class _Ty0, class _Alloc, enable_if_t<!is_unbounded_array_v<_Ty0>, int> = 0>
    friend shared_ptr<_Ty0> allocate_shared_for_overwrite(
        const _Alloc& _Al_arg);

    template <class _Ty0, enable_if_t<is_unbounded_array_v<_Ty0>, int> = 0>
    friend shared_ptr<_Ty0> make_shared_for_overwrite(size_t _Count);

    template <class _Ty0, class _Alloc, enable_if_t<is_unbounded_array_v<_Ty0>, int> = 0>
    friend shared_ptr<_Ty0> allocate_shared_for_overwrite(
        const _Alloc& _Al_arg, size_t _Count);

```

Reply:
It seemed a bit weird to me too, but I tried to copy the style used in `make_shared`.

Reply:
This was my doing in #309, where I took @AdamBucior's elegant concepts-based code and reimplemented it with ~~neolithic stone tools~~ SFINAE, because not all of our supported compilers have implemented concepts yet (e.g. IntelliSense). While we ordinarily use SFINAE in a highly consistent manner as @miscco noted, it doesn't work here :crying_cat_face: due to the friendship (plus interaction with compiler bugs, IIRC). I think I was a bad kitty - instead of reducing and reporting the compiler bug, I perma-worked-around it (since default arg SFINAE would have needed verbose forward declarations anyways) by using ~~paleolithic stone tools~~ return type SFINAE.

(The issue with template argument SFINAE was whether friends can introduce default template arguments.)

I doubled down on my bad kitty behavior :smirk_cat: by ~~jumping up on the table where I'm not allowed~~ not marking this workaround as TRANSITION, because the SFINAE is ultimately doomed. In the future, when our C++20 compilers all support concepts, we can eradicate this and go back to Adam's original, elegant code.